### PR TITLE
[WIP][Core] Optional Profiling.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Screen Shot 2020-07-21 at 3 08 07 PM

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by simon-mo** [comment]
slightly prefer the name `enable_profile` to be more consistent with other feature flags.

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -329,3 +329,7 @@ RAY_CONFIG(bool, put_small_object_in_memory_store, false)
 /// the owner has been granted a lease. A value >1 is used when we want to enable
 /// pipelining task submission.
 RAY_CONFIG(uint32_t, max_tasks_in_flight_per_worker, 1)
+
+RAY_CONFIG(bool, run_profiling,
+           getenv("RAY_PROFILE_ENABLED") != nullptr &&
+               getenv("RAY_PROFILE_ENABLED") == std::string("true"))
```
Comment (simon-mo) [inline src/ray/common/ray_config_def.h:335]: We probably shouldn't feature flag it via environment variable? Seems to me just enable it via _internal_config is better.

Comment (rkooo567) [reply src/ray/common/ray_config_def.h:335]: One concern I have is that `_internal_config` (not sure if it is still the case) is not meant to be exposed to users, but this feature is supposed to be exposed to users.

**Comment by ericl** [comment]
Why don't we leave profiling on by default for now? Actually looking at the delta, 5-25% performance hit is pretty reasonable to get monitoring. Performance sensitive applications can turn it of to get the extra tiny bit of speed.

**Comment by rkooo567** [comment]
>Why don't we leave profiling on by default for now

For now means until we switch this to use new metrics infra internally? 

I am honestly not sure if this is counted as monitoring though. All profiling data is exposed to users only through ray.timeline, which I believe is extremely rarely used. It is not used by dashboard or any other internal stuff as far as I know. (Let me know if it is wrong. If so, I am totally on the same page to turn it on by default). 

I understand the usefulness of ray.timeline (because I used before), but It looks more like unnecessary overhead for 99% of user scenarios.

**Comment by rkooo567** [comment]
@ericl @edoakes @simon-mo @robertnishihara 

I failed to reproduce 20~50% performance improvement which I could get from my local Mac machine. Since we only have a little bit of performance improvement, I will just close this PR. Instead I will spend some time in the future (maybe next Q) to share new metrics infrastructure to measure profiling data.

**Pull request closed without merging**